### PR TITLE
No echo and output buffering in App::run()

### PR DESCRIPTION
**Pull request by mvorisek**
No phpunit warnings:
image

### CODE REVIEW

**Review by georgehristov — APPROVED** [review]
LGTM

File: src/App.php [hunk 1]
```diff
@@ -516,12 +515,12 @@ public function run()
             if (isset($_GET['__atk_callback']) && $this->catch_runaway_callbacks) {
                 throw new Exception('Callback requested, but never reached. You may be missing some arguments in request URL.');
             }
-            echo $this->html->template->render();
+
+            $output = $this->html->template->render();
```
Comment (georgehristov) [inline src/App.php:519]: Probably good to initialize `$output` for cases this line is not reached

**Review by georgehristov — APPROVED** [review]

File: src/App.php (hunk 1 above)
Comment (mvorisek) [reply src/App.php:519]: You are right! But still working on other unclosed buffers, see warnings in unit tests.

File: src/App.php [hunk 2]
```diff
@@ -1012,8 +1012,6 @@ function () {
                 if (!$this->run_called) {
                     try {
                         $this->run();
-                    } catch (ExitApplicationException $e) {
```
Comment (mvorisek) [inline src/App.php:1015]: catched in `App::run()`, @georgehristov @abbadon1334 @ibelar wdyt?

File: tests/DemosTest.php [hunk 3]
```diff
@@ -144,8 +144,7 @@ protected function getClient(): Client
                     throw $e;
                 }
             } finally {
-                $body = ob_get_contents();
-                ob_end_clean();
+                $body = ob_get_clean();
```
Comment (mvorisek) [inline tests/DemosTest.php:167]: the same, one line only...

**Pull request merged**